src/python_framework_v02/troute/nhd_io.py

```diff
@@ -527,3 +528,69 @@ def get_reservoir_restart_from_wrf_hydro(
     init_waterbody_states = mod
 
     return init_waterbody_states
+
+
+def build_last_obs_df(routlink_file, wrf_last_obs_flag, fvd_df):
+    # open routelink_file and extract discharges
+    with xr.open_dataset(routlink_file) as ds:
+        df_model_discharges = ds["model_discharge"].to_dataframe()
+        df_discharges = ds["discharge"].to_dataframe()
+        last_ts = df_model_discharges.index.get_level_values("timeInd")[-1]
+        model_discharge_last_ts = df_model_discharges[
+            df_model_discharges.index.get_level_values("timeInd") == last_ts
+        ]
+        discharge_last_ts = df_discharges[
+            df_discharges.index.get_level_values("timeInd") == last_ts
+        ]
+        df1 = ds["stationId"].to_dataframe()
+        df1 = df1.astype(int)
+        model_discharge_last_ts = model_discharge_last_ts.join(df1)
+        model_discharge_last_ts = model_discharge_last_ts.join(discharge_last_ts)
+        model_discharge_last_ts = model_discharge_last_ts.loc[
+            model_discharge_last_ts["model_discharge"] != -9999.0
+        ]
+        model_discharge_last_ts = model_discharge_last_ts.reset_index().set_index(
+            "stationId"
+        )
+        model_discharge_last_ts = model_discharge_last_ts.drop(
+            ["stationIdInd", "timeInd"], axis=1
+        )
+        # If predict from last_obs file use last obs file results
+        if wrf_last_obs_flag == True:
+            model_discharge_last_ts["last_nudge"] = (
+                model_discharge_last_ts["discharge"]
+                - model_discharge_last_ts["model_discharge"]
+            )
+        # Else predict from the model outputs from t-route if index doesn't match interrupt computation as the results won't be valid
+        else:
+            fvd_df = fvd_df
+            if len(model_discharge_last_ts.index) == len(fvd_df.index):
+                model_discharge_last_ts["last_nudge"] = (
+                    model_discharge_last_ts["discharge"] - fvd_df[fvd_df.columns[0]]
+                )
+            else:
+                print("THE NUDGING FILE IDS DO NOT MATCH THE FLOWVELDEPTH IDS")
+                sys.exit()
+        # Predictions created with continuously decreasing deltas until near 0 difference
+        a = 120
+        prediction_df = pd.DataFrame(index=model_discharge_last_ts.index)
+
+        for time in range(0, 720, 5):
+            weight = np.exp(time / -120)
```

Comment:
the weight is eventually applied to an array and it might be efficient to stack the computation into the array operation and perhaps take advantage of numpy optimizations. Just thinking...